infra/src/main/kotlin/fi/fta/geoviite/infra/inframodel/InfraModelParsing.kt

```diff
@@ -176,14 +176,28 @@ val xmlCharsets = listOf(
     StandardCharsets.US_ASCII,
     StandardCharsets.ISO_8859_1,
 )
-fun getEncoding(bytes: ByteArray): Charset {
-    ByteArrayInputStream(bytes).use { stream ->
-        val xmlStreamReader: XMLStreamReader = XMLInputFactory.newInstance().createXMLStreamReader(stream)
-        val fileEncoding = xmlStreamReader.encoding
-        val encodingFromXMLDeclaration = xmlStreamReader.characterEncodingScheme
-        return (encodingFromXMLDeclaration ?: fileEncoding)?.let { name ->
+
+fun mapBomToCharset(bom: ByteOrderMark) =
+    when (bom) {
+        ByteOrderMark.UTF_8 -> StandardCharsets.UTF_8
+        ByteOrderMark.UTF_16BE -> StandardCharsets.UTF_16BE
+        ByteOrderMark.UTF_16LE -> StandardCharsets.UTF_16LE
+        else -> null
+    }
+
+fun encodingsFromXmlStream(stream: InputStream) =
+    XMLInputFactory.newInstance().createXMLStreamReader(stream).let {
+        xmlStreamReader -> xmlStreamReader.encoding to xmlStreamReader.characterEncodingScheme
+    }
+
+fun getEncodingAndBom(bytes: ByteArray): Pair<Charset, Boolean> {
+    BOMInputStream(ByteArrayInputStream(bytes)).use { stream ->
+        val encodingFromBOM = stream.bom?.let { bom -> mapBomToCharset(bom) }?.name()
+        val (fileEncoding, encodingFromXMLDeclaration) = encodingsFromXmlStream(stream)
+        val encoding = (encodingFromBOM ?: encodingFromXMLDeclaration ?: fileEncoding)?.let { name ->
```

Comment:
Preferoidaan `BOMInputStream`:in palauttama enkoodaustieto kaiken muun yli. Tiedoston enkoodaus voi vaihtua vaikka esim. jonkun sopivasti ristiin menneen tallennuksen takia, ja tällöin XML:n declaration voi jäädä osoittamaan vanhaan merkistöön. Entinen koodi heitti tilanteissa, joissa tiedosto väitti olevansa jotain muuta kuin UTF:ää, mutta sen alusta löytyi minkä tahansa UTF:n BOM.

Reply:
Näyttää hyvältä :+1: